syntax = "proto3";

package post;

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  int32 creator_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message UpdatePostRequest {
  int32 id = 1;
  string title = 2;
  string description = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message DeletePostRequest {
  int32 id = 1;
  int32 creator_id = 2;
}

message GetPostRequest {
  int32 id = 1;
  int32 requester_id = 2;
}

message ListPostsRequest {
  int32 creator_id = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message PostResponse {
  Post post = 1;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total = 2;
}

message DeletePostResponse {
  bool success = 1;
}

service PostService {
  rpc CreatePost(CreatePostRequest) returns (PostResponse);
  rpc UpdatePost(UpdatePostRequest) returns (PostResponse);
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse);
  rpc GetPost(GetPostRequest) returns (PostResponse);
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
}
